Events should bubble up the tree

abarth opened the pull request:
Previously we reversed the event path by mistake before dispatching events.

abarth commented:
@Hixie @chinmaygarde

abarth commented:
@afandria

Hixie commented:
Can you elaborate on "Previously we reversed the event path by mistake before dispatching events"?

abarth commented:
I thought we constructed the event path starting at the top of the tree (as the docs used to indicate).  That meant we needed to reverse the path when dispatching events.  However, it turns out we construct the event path from the leaves towards the root (as we unwind the recursion), so we don't need to reverse the path when dispatching.  This patch updates the docs to match reality and then updates the event dispatch algorithm to dispatch from the leaves up.

Hixie commented:
so when did the code break? didn't this work before?

abarth commented:
It probably broke when I changed how hit testing worked.  Previously, it was the parent's responsibility to bounds-test the children, but that didn't work for things like RenderTransform that paint in a location that the parent didn't understand.  I moved the bounds test into the child and probably reversed the order in which we create the path by mistake.

Hixie commented:
ah, that would make sense. Fair enough. LGTM.
